[RNMobile] `BlockDraggable` component

fluiddot opened the pull request:
Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4620.

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Add `BlockDraggable` component for enabling dragging blocks.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

This component is required for providing the drag & drop functionality for blocks, similarly as we have in the web version.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Based on the logic of the same component on the web version, this component handles the following functionality:
- Uses the `Draggable` block to determine when a block is being dragged, as well as finding via the blocks layouts the `clientId` of the block.
- Displays a chip component that moves along with the dragging gesture.
- Collapses the block being dragged to help the user visualize the drop target.
- When the dragged block is out of the scroll view when the dragging gesture begins, it automatically scrolls the block list to the origin position of the block. This will help the user to know where the block is originally placed.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
### Drag a block
1. Open a post/page.
2. Add some blocks.
3. Tap over a block for around one second.
4. Observe that the block collapses and a visual clue is shown on top of the touch position.
5. Observe that the visual clue displays the icon of the block being dragged.
6. Move the finger around the screen and observe that the visual clue follows the dragging gesture.
7. Release the finger from the screen and observe that the block expands.

### Drag a block partially off of scroll view
1. Open a post/page.
2. Add some blocks.
3. Add a Gallery block with several items.
4. Scroll the block list until part of the Gallery block is off the scroll view.
5. Tap over the Gallery block for around one second.
6. Observe that the block collapses and a visual clue is shown on top of the touch position.
7. Observe that the block list is scrolled to the position where the Gallery block is placed.
8. Observe that the visual clue displays the icon of the block being dragged.
9. Move the finger around the screen and observe that the visual clue follows the dragging gesture.
10. Release the finger from the screen and observe that the block expands.

**NOTES:** Sometimes text blocks might interfere with the long-press gesture. I haven't found yet a solution for this, but I'd like to note that this should be addressed before releasing the drag&drop feature.

### HTML editor
Due to a recent conflict with the gesture handler library, the HTML editor has been updated hence, it would be nice to check that it works as expected.
1. Open a post/page.
2. Add some blocks.
3. Tap on the three dots button to open the post settings.
4. Tap on "Switch to HTML".
5. Observe that the HTML version of the post/page content is displayed.
6. Observe that the content can be scrolled.
7. Observe that the content can be edited.

## Screenshots or screencast <!-- if applicable -->

**Light mode:**
iOS|Android
--|--
<video src=https://user-images.githubusercontent.com/14905380/160593878-f944644c-7fe0-45de-89d3-8292b53a7607.MP4>|<video src=https://user-images.githubusercontent.com/14905380/160593932-46759e4b-8818-4abb-a868-700e37589bd0.mp4>

**Dark mode:**
iOS|Android
--|--
<video src=https://user-images.githubusercontent.com/14905380/160594012-36c93b71-962e-49e3-947d-a04c2d577bc8.MP4>|<video src=https://user-images.githubusercontent.com/14905380/160594048-b3535007-c09a-4565-add4-060fc8fa955a.mp4>

fluiddot committed:
Use animated scroll handler in KeyboardAwareFlatList

fluiddot committed:
Add hook for scrolling the block list while dragging

fluiddot committed:
Improve scroll animation in useScrollWhenDragging

fluiddot committed:
Add draggable chip component

fluiddot committed:
Add block draggable component

fluiddot committed:
Remove icon prop from draggable chip component

fluiddot committed:
Add draggable placeholder

fluiddot committed:
Fix draggable chip location

fluiddot committed:
Wrap BlockListItemCell with BlockDraggable

fluiddot committed:
Fix block draggable placeholder style

fluiddot committed:
Animate scale property instead of opacity of draggable chip

fluiddot committed:
Fix draggable placeholder container height calculation

fluiddot committed:
Fix BlockDraggable height animation

fluiddot committed:
Move draggable to BlockDraggableWrapper

fluiddot committed:
Disable isDragging when long-press gesture ends

fluiddot committed:
Fix onLayout calculation in block list item cell

fluiddot committed:
Add findBlockLayoutByPosition helper

fluiddot committed:
Set up dragging block by position

fluiddot committed:
Remove animate scroll velocity

fluiddot committed:
Remove useScrollWhenDragging hook

This hook will be introduced in a separate PR

fluiddot committed:
Remove react-native-reanimated mock

fluiddot commented:
The `React Native E2E Tests (iOS)` PR check is failing because on iOS the HTML mode is not working. Specifically, the text input of the HTML content doesn't allow scrolling. I haven't found yet the root cause of this issue but most likely it's related to some kind of conflict between `react-native-gesture-handler` and RN's `PanResponder`.

fluiddot commented on packages/block-editor/src/components/block-draggable/draggable-chip.native.js:
```diff
@@ -0,0 +1,51 @@
+import styles from './style.scss';
+import { store as blockEditorStore } from '../../store';
+
+const shadowStyle = {
+	shadowColor: '#000',
+	shadowOffset: {
+		width: 0,
+		height: 2,
+	},
+	shadowOpacity: 0.25,
+	shadowRadius: 3.84,
+
+	elevation: 5,
+};
```
I defined the shadow style properties here because these properties are not recognized in SCSS.

fluiddot commented on packages/block-editor/src/components/block-draggable/draggable-chip.native.js:
```diff
@@ -0,0 +1,51 @@
+};
+
+export default function BlockDraggableChip() {
+	const { blockIcon } = useSelect( ( select ) => {
+		const { getBlockName, getDraggedBlockClientIds } = select(
+			blockEditorStore
+		);
+		const draggedBlockClientIds = getDraggedBlockClientIds();
+		const blockName = getBlockName( draggedBlockClientIds[ 0 ] );
+
+		return {
+			blockIcon: getBlockType( blockName )?.icon,
+		};
+	} );
```
I thought of passing the `blockIcon` from the `BlockDraggable` component (i.e. its parent component) but I experienced downgrades in the performance due to re-renders caused by Redux state updates. For this reason, I finally decided to retrieve the `blockIcon` directly here.

fluiddot commented on packages/block-editor/src/components/block-draggable/style.native.scss:
```diff
@@ -0,0 +1,17 @@
+	padding: 16px;
+	background-color: #f7f7f7;
+	border-radius: 8px;
+}
```
I tried to follow a similar design to this example:

![Untitled](https://user-images.githubusercontent.com/14905380/159916849-3c96dac0-9558-449e-9671-fee7b098a648.png)

fluiddot commented on packages/block-editor/src/components/block-list/block-list-context.native.js:
```diff
@@ -7,12 +7,25 @@ export const DEFAULT_BLOCK_LIST_CONTEXT = {
 
+function findBlockLayoutByPosition( data, position ) {
+	// Only enabled for root level blocks
+	return Object.values( data ).find( ( block ) => {
+		return (
+			position.x >= block.x &&
+			position.x <= block.x + block.width &&
+			position.y >= block.y &&
+			position.y <= block.y + block.height
+		);
+	} );
```
In the future, we could support inner blocks by traversing the blocks layouts. Although, we should keep in mind that might be difficult for the user to drag a container block (e.g. a Gallery block) vs the inner blocks (e.g. Gallery items). One approach to address this would be to only allow dragging the inner blocks when the block is selected.

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -45,6 +45,7 @@ export default function Draggable( {
 		} )
 		.onEnd( () => {
 			'worklet';
+			isDragging.value = false;
```
Looks like we overlooked setting `isDragging` to `false`, this was making the draggable logic to working incorrectly.

fluiddot commented on test/native/setup.js:
```diff
@@ -153,16 +153,6 @@ jest.mock( '@react-native-community/blur', () => () => 'BlurView', {
 	virtual: true,
 } );
 
-jest.mock( 'react-native-reanimated', () => {
-	const Reanimated = require( 'react-native-reanimated/mock' );
-
-	// The mock for `call` immediately calls the callback which is incorrect
-	// So we override it with a no-op
-	Reanimated.default.call = () => {};
-
-	return Reanimated;
-} );
```
This mock is no longer required as Reanimated provides its own configuration for Jest ([reference](https://github.com/WordPress/gutenberg/blob/2bd4e6696522340eede0690c2f413ab7b1ff3b51/test/native/setup.js#L16)).

fluiddot commented on packages/components/src/mobile/keyboard-aware-flat-list/index.android.js:
```diff
@@ -2,17 +2,27 @@
+	const scrollHandler = useAnimatedScrollHandler( { onScroll } );
+	return (
+		<KeyboardAvoidingView style={ { flex: 1 } }>
+			<AnimatedFlatList
+				ref={ innerRef }
+				onScroll={ scrollHandler }
+				{ ...props }
+			/>
```
We need to use the animated version of the `FlatList` component in order to animate the scroll.

fluiddot commented on packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js:
```diff
@@ -19,53 +26,64 @@ export const KeyboardAwareFlatList = ( {
-		} }
-		scrollEnabled={ listProps.scrollEnabled }
-		onScroll={ ( event ) => {
-			this.latestContentOffsetY = event.nativeEvent.contentOffset.y;
```
This part is now done in the `scrollHandler` created by `useAnimatedScrollHandler`.

fluiddot commented on packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js:
```diff
@@ -4,13 +4,20 @@
 import { memo } from '@wordpress/element';
 
 const List = memo( FlatList, isEqual );
+const AnimatedKeyboardAwareScrollView = Animated.createAnimatedComponent(
+	KeyboardAwareScrollView
+);
```
We need to use the animated version of the `KeyboardAwareScrollView` component in order to animate the scroll.

fluiddot commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -310,6 +313,7 @@ export class BlockList extends Component {
 					ListHeaderComponent={ header }
 					ListEmptyComponent={ ! isReadOnly && this.renderEmptyList }
 					ListFooterComponent={ this.renderBlockListFooter }
+					onScroll={ onScroll }
```
We need to listen to the scroll event for different calculations on the `BlockDraggable` component.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,300 @@
+		scrollRef,
+		findBlockLayoutByPosition,
+	} = useBlockListContext();
+	const animatedScrollRef = useAnimatedRef();
+	animatedScrollRef( scrollRef );
```
In order to use [Reanimated's `scrollTo` function](https://docs.swmansion.com/react-native-reanimated/docs/api/nativeMethods/scrollTo), we have to get the animated ref of the scroll. Since `scrollRef` is already provided via the block list context, we create the animated ref by invoking the function that sets the reference ([source code reference](https://github.com/software-mansion/react-native-reanimated/blob/45c57a906b1c3881c37f08b09749a9c68faf7ce1/src/reanimated2/hook/useAnimatedRef.ts#L26)).

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,300 @@
+
+		const foundClientId = blockLayout?.clientId;
+		if ( foundClientId ) {
+			startDraggingBlocks( [ foundClientId ] );
```
Calling this Redux action will notify the dragged block that is being dragged.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,300 @@
+		'worklet';
+		const scrollLayout = measure( animatedScrollRef );
+		scroll.x.value = scrollLayout.pageX;
+		scroll.y.value = scrollLayout.pageY;
```
Since we use absolute positions (i.e. positions relative to the window), we need to know the position of the scroll for the position calculations like moving the chip.

fluiddot committed:
Rename CHIP_OFFSET_TO_TOUCH_POSITION constant

fluiddot committed:
Remove unused shared values of chip component

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,300 @@
+		} else {
+			runOnUI( stopBlockDragging )();
+		}
+	}, [ isBeingDragged ] );
```
The animations applied to the block are triggered when the block detects that is being dragged.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,300 @@
+				containerHeightBeforeDragging.value === 0 ||
+				collapseAnimation.value === 0
+					? 'auto'
+					: height,
```
We only set a fixed height when the block is collapsed or when the container height value hasn't been calculated (i.e. value is `0`).

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,300 @@
+
+	const placeholderStyles = useAnimatedStyle( () => {
+		return {
+			display: collapseAnimation.value === 0 ? 'none' : 'flex',
```
We only show the placeholder component when the block is collapsed.

fluiddot commented:
@geriux The `React Native E2E Tests (iOS)` PR check is failing because on iOS the HTML mode is not working. Specifically, the text input of the HTML content doesn't allow scrolling. I haven't found yet the root cause of this issue but most likely it's related to some kind of conflict between `react-native-gesture-handler` and RN's `PanResponder`.

fluiddot committed:
Stop dragging when no block is found

fluiddot committed:
Fix drag position calculation

fluiddot committed:
Update html text input styles

fluiddot committed:
Unify container component within html text input

fluiddot commented:
> @geriux The `React Native E2E Tests (iOS)` PR check is failing because on iOS the HTML mode is not working. Specifically, the text input of the HTML content doesn't allow scrolling. I haven't found yet the root cause of this issue but most likely it's related to some kind of conflict between `react-native-gesture-handler` and RN's `PanResponder`.

This issue has been addressed in https://github.com/WordPress/gutenberg/pull/39617/commits/5c6e21acde9789052eef5992d8f7b671ae8c70dc 🎊 .

fluiddot committed:
Use only a single client id in block draggable

fluiddot committed:
Add documentation to block draggable components

fluiddot committed:
Add documentation to block draggable chip component

fluiddot committed:
Add documentation to findBlockLayoutByPosition

fluiddot committed:
Update scrollOffsetTarget calculation

fluiddot committed:
Fix typos in block draggable

fluiddot committed:
Add draggable wrapper container style

fluiddot commented on packages/components/src/mobile/html-text-input/index.native.js:
```diff
@@ -90,32 +96,42 @@ export class HTMLTextInput extends Component {
 			...( style?.text && { color: style.text } ),
 		};
 		return (
-			<HTMLInputContainer parentHeight={ this.props.parentHeight }>
-				<TextInput
-					autoCorrect={ false }
-					accessibilityLabel="html-view-title"
-					textAlignVertical="center"
-					numberOfLines={ 1 }
-					style={ titleStyle }
-					value={ this.props.title }
-					placeholder={ __( 'Add title' ) }
-					placeholderTextColor={ placeholderStyle.color }
-					onChangeText={ this.props.editTitle }
-				/>
-				<TextInput
-					autoCorrect={ false }
-					accessibilityLabel="html-view-content"
-					textAlignVertical="top"
-					multiline
-					style={ htmlStyle }
-					value={ this.state.value }
-					onChangeText={ this.edit }
-					onBlur={ this.stopEditing }
-					placeholder={ __( 'Start writing…' ) }
-					placeholderTextColor={ placeholderStyle.color }
-					scrollEnabled={ HTMLInputContainer.scrollEnabled }
-				/>
-			</HTMLInputContainer>
```
The `HTMLInputContainer` component has been removed because we no longer need to provide a platform-specific implementation for rendering the HTML input. Instead, now we'll use the implementation previously used for Android ([reference](https://github.com/WordPress/gutenberg/blob/4e5f5f7f00264762b2f612992e710c629460fbc6/packages/components/src/mobile/html-text-input/container.android.js)).

geriux replied:
These changes look good to me, even though we needed this change for the incompatibility, I think it makes sense to unify it. I tested it both on iOS and Android.

fluiddot commented on packages/components/src/mobile/html-text-input/index.native.js:
```diff
@@ -90,32 +96,42 @@ export class HTMLTextInput extends Component {
+						placeholder={ __( 'Start writing…' ) }
+						placeholderTextColor={ placeholderStyle.color }
+						scrollEnabled={ false }
+						// [Only iOS] This prop prevents the text input from
+						// automatically getting focused after scrolling
+						// content.
+						rejectResponderTermination={ false }
```
This prop is used for addressing the issue that originally was fixed by using `PanResponder` in `KeyboardAvoidingView` component on iOS.

geriux replied:
TIL!

fluiddot commented on packages/components/src/mobile/html-text-input/container.ios.js:
```diff
@@ -1,50 +0,0 @@
-	constructor() {
-		super( ...arguments );
-
-		this.panResponder = PanResponder.create( {
-			onStartShouldSetPanResponderCapture: () => true,
-
-			onPanResponderMove: ( e, gestureState ) => {
-				if ( gestureState.dy > 100 && gestureState.dy < 110 ) {
-					// Keyboard.dismiss() and this.textInput.blur() are not working here
-					// They require to know the currentlyFocusedID under the hood but
-					// during this gesture there's no currentlyFocusedID.
-					UIManager.blur( e.target );
-				}
-			},
-		} );
```
AFAIK this logic was used as a workaround for preventing the `TextInput` component from automatically gaining focus when releasing the touch after scrolling. However, this solution no longer works due to the use of the gesture handler library via the `GestureDetector` component ([reference](https://github.com/WordPress/gutenberg/blob/4e5f5f7f00264762b2f612992e710c629460fbc6/packages/components/src/draggable/index.native.js#L77)).

geriux commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,315 @@
+	}
+
+	return (
+		<Animated.View onLayout={ onContainerLayout } style={ containerStyles }>
```
I'm wondering if we could try to get this data from the `blocksLayouts`. The size should be the same and it'd avoid having `onLayout` twice per block. What do you think?

fluiddot replied:
Good point, yep, I think we could save this `onLayout` listener by using the blocks layout data provided by the block list context 👍 . I'm going to do a quick check to verify that we receive the same values and if so, replace it with that data.

fluiddot replied:
I verified that we get the same value from the blocks layout data as `onLayout`, so I'll proceed to update this.

fluiddot replied:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/39617/commits/d16ce157ebc7b6e916c9c3026da48a525a71ca1b.

fluiddot replied:
Heads up that I had to update the collapse/expand animation due to these changes in https://github.com/WordPress/gutenberg/pull/39617/commits/bda7b248e6a583f497f75fd4578699c6db9a85f3. You can watch the new animation in the videos referenced in the PR's description.

geriux commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -0,0 +1,315 @@
+				{ children( { isDraggable: true } ) }
+			</Animated.View>
+			<Animated.View
+				style={ [
```
Would it make sense to have the styles composition in a variable outside of the render?

fluiddot replied:
Originally I thought of providing all the styles for this component via `useAnimatedStyle` but Reanimated recommends keeping dynamic and static styles separated.

[Extracted from Reanimated's `useAnimatedStyle` documentation](https://docs.swmansion.com/react-native-reanimated/docs/api/hooks/useAnimatedStyle#returns):
> We recommend that you defined static styles using React Native's [StyleSheet API](https://reactnative.dev/docs/stylesheet) and pass them along the animated styles object returned by this hook. It is more efficient to only keep styles that are actually changed as a result of animation in the animated style hook, and the rest of the styles should be either provided as inline objects or using StyleSheets.

Being that said, I could create another constant with both, something like:
```
const placeholderDynamicStyles = useAnimatedStyle( () => {
	return {
		display: collapseAnimation.value === 0 ? 'none' : 'flex',
		opacity: collapseAnimation.value,
	};
} );
const placeholderStyles = [
	styles[ 'draggable-placeholder__container' ],
	placeholderDynamicStyles,
];
```

Wdyt?

geriux replied:
> Being that said, I could create another constant with both, something like:

That sounds good, let's do that

fluiddot replied:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/39617/commits/4060e1ad86e6d2c9b880a7ed28509668c146f062.

geriux commented on packages/block-editor/src/components/block-draggable/style.native.scss:
```diff
@@ -0,0 +1,21 @@
+.draggable-chip__container {
+	flex-direction: row;
+	padding: 16px;
+	background-color: #f7f7f7;
```
I feel like this could match `$gray-0` no?

fluiddot replied:
Good catch! Yeah, I rather use a predefined color so I'll use `$gray-0`, thanks for the suggestion 🙇 .

fluiddot replied:
Now that you mention that, I noticed that I forgot to add the dark mode styles 😅, I'll add them as soon as possible.

geriux replied:
>I noticed that I forgot to add the dark mode styles 😅, I'll add them as soon as possible.

That's true! 😄 thanks!

fluiddot replied:
I've just applied this suggestion and added dark mode styles in the following commits:
- [Add dark mode styles for draggable chip](https://github.com/WordPress/gutenberg/pull/39617/commits/e9ce4e3e52b681385d37669ce24448ea0868e869)
- [Add dark mode styles for block draggable](https://github.com/WordPress/gutenberg/pull/39617/commits/4060e1ad86e6d2c9b880a7ed28509668c146f062)

geriux commented on packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js:
```diff
@@ -19,53 +26,64 @@ export const KeyboardAwareFlatList = ( {
+			onKeyboardWillHide={ () => {
+				this.keyboardWillShowIndicator = false;
+			} }
+			onKeyboardDidHide={ () => {
```
Not relevant for this PR but we should update this code to avoid using that many inline functions 😅

fluiddot replied:
I totally agree, we have quite a lot of callbacks as inline functions, which might even impact the performance. I think I could spare some time to address this in this PR.

fluiddot replied:
I replaced the inline functions in https://github.com/WordPress/gutenberg/pull/39617/commits/842b09b296409c99705ed5a09972ce81da02a4f5.

geriux reviewed (commented):
This is looking great! I've tested this while working on the other tasks and it's working as expected, I've left some minor comments, let me know what you think.

fluiddot committed:
Add dark mode styles for draggable chip

fluiddot committed:
Add dark mode styles for block draggable

fluiddot committed:
Get container height from blocks layout data

fluiddot committed:
Replace inline callback functions with useCallback hook

fluiddot committed:
Update collapse/expand animation when dragging a block

fluiddot commented:
> This is looking great! I've tested this while working on the other tasks and it's working as expected, I've left some minor comments, let me know what you think.

@geriux I pushed some changes for addressing the feedback in previous comments, so the PR is ready for another review, thanks 🙇 .

**NOTE:** I've also updated the PR's description to include videos using the dark mode.

geriux commented on packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js:
```diff
@@ -29,14 +29,54 @@ export const KeyboardAwareFlatList = ( {
+			scrollViewRef.current = ref;
+			innerRef( ref );
+		},
+		[ scrollViewRef, innerRef ]
```
I wonder if this should have the `.current` value as a dependency instead?

fluiddot replied:
I think that's not really necessary because we're accessing the reference (i.e. `scrollViewRef.current`). Although the function is memoized, while `scrollViewRef doesn't change, the reference retrieved by `current` property will have the latest value.

Actually, I'm thinking that most likely we even don't need to add the references (i.e. `scrollViewRef`) in the dependency list, since [React guarantees their inmutability](https://reactjs.org/docs/hooks-reference.html#useref). I'll remove `scrollViewRef` from here and do the same on the rest of the callbacks.

fluiddot replied:
I've applied the change referenced in the [previous comment](https://github.com/WordPress/gutenberg/pull/39617#discussion_r837530281) in https://github.com/WordPress/gutenberg/pull/39617/commits/d66fb3d17ec2a4cf8bd2a4006f9604f6b96efe10

fluiddot committed:
Force draggable chip to be displayed upfront

fluiddot committed:
Remove refs from dependencies arrays

References can be omitted from the dependencies arrays since React guarantees that they are inmutable.

geriux reviewed (approved):
LGTM! 👏 I tested this on both iOS and Android, light/dark mode changes look good as well. Nice work!
